plugins/ja3_fingerprint/ja3_fingerprint.cc

```diff
@@ -163,112 +178,111 @@ append_to_field(TSMBuffer bufp, TSMLoc hdr_loc, const char *field, int field_len
+  TSVConn const ssl_vc{static_cast<TSVConn>(edata)};
+  delete static_cast<ja3_data *>(TSUserArgGet(ssl_vc, ja3_idx));
+  TSUserArgSet(ssl_vc, ja3_idx, nullptr);
+  TSVConnReenable(ssl_vc);
+  return TS_SUCCESS;
+}
 
-    // Validate that the buffer is the same size as we will be writing into (compile time)
-    static_assert((16 * 2 + 1) == sizeof(data->md5_string));
-    for (int i = 0; i < 16; i++) {
-      snprintf(&(data->md5_string[i * 2]), sizeof(data->md5_string) - (i * 2), "%02x", static_cast<unsigned int>(digest[i]));
-    }
-    Dbg(dbg_ctl, "Fingerprint: %s", data->md5_string);
-    break;
+static void
+modify_ja3_headers(TSCont contp, TSHttpTxn txnp, ja3_data const *ja3_vconn_data)
+{
+  // Decide global or remap
+  ja3_remap_info *remap_info = static_cast<ja3_remap_info *>(TSContDataGet(contp));
+  bool            raw_flag   = remap_info ? remap_info->raw_enabled : global_raw_enabled;
+  bool            log_flag   = remap_info ? remap_info->log_enabled : global_log_enabled;
+  Dbg(dbg_ctl, "Found ja3 string.");
+
+  // Get handle to headers
+  TSMBuffer bufp;
+  TSMLoc    hdr_loc;
+  if (global_modify_incoming_enabled) {
+    TSAssert(TS_SUCCESS == TSHttpTxnClientReqGet(txnp, &bufp, &hdr_loc));
+  } else {
+    TSAssert(TS_SUCCESS == TSHttpTxnServerReqGet(txnp, &bufp, &hdr_loc));
   }
-  case TS_EVENT_VCONN_CLOSE: {
-    // Clean up
-    ja3_data *data = static_cast<ja3_data *>(TSUserArgGet(ssl_vc, ja3_idx));
-
-    if (data == nullptr) {
-      Dbg(dbg_ctl, "client_hello_ja3_handler(): Failed to retrieve ja3 data at VCONN_CLOSE.");
-      return TS_ERROR;
-    }
 
-    TSUserArgSet(ssl_vc, ja3_idx, nullptr);
+  // Add JA3 md5 fingerprints
+  append_to_field(bufp, hdr_loc, "X-JA3-Sig", 9, ja3_vconn_data->md5_string, 32);
 
-    delete data;
-    break;
-  }
-  default: {
-    Dbg(dbg_ctl, "client_hello_ja3_handler(): Unexpected event.");
-    break;
+  // If raw string is configured, added JA3 raw string to header as well
+  if (raw_flag) {
+    append_to_field(bufp, hdr_loc, "x-JA3-RAW", 9, ja3_vconn_data->ja3_string.data(), ja3_vconn_data->ja3_string.size());
   }
+  TSHandleMLocRelease(bufp, TS_NULL_MLOC, hdr_loc);
+
+  // Write to logfile
+  if (log_flag) {
+    TSTextLogObjectWrite(pluginlog, "Client IP: %s\tJA3: %.*s\tMD5: %.*s", ja3_vconn_data->ip_addr,
+                         static_cast<int>(ja3_vconn_data->ja3_string.size()), ja3_vconn_data->ja3_string.data(), 32,
+                         ja3_vconn_data->md5_string);
   }
-  TSVConnReenable(ssl_vc);
-  return TS_SUCCESS;
 }
 
 static int
 req_hdr_ja3_handler(TSCont contp, TSEvent event, void *edata)
 {
   TSEvent expected_event = global_modify_incoming_enabled ? TS_EVENT_HTTP_READ_REQUEST_HDR : TS_EVENT_HTTP_SEND_REQUEST_HDR;
   if (event != expected_event) {
-    TSError("[%s] req_hdr_ja3_handler(): Unexpected event, got %d, expected %d", PLUGIN_NAME, event, expected_event);
+    TSError("Unexpected event, got %d, expected %d", event, expected_event);
```

Comment:
We do need to keep the `PLUGIN_NAME` in the TSError messages, however. Otherwise it's not trivial to figure out that the message is related to the plugin in the set of other diags.log messages.